code/datums/client_image_group.dm

```diff
@@ -0,0 +1,90 @@
+var/global/list/datum/client_image_group/client_image_groups
+
+/datum/client_image_group
+	var/list/image/images
+	var/list/loc_to_image_lookup
+	var/list/subscribed_mobs
+
+	New()
+		. = ..()
+		images = list()
+		loc_to_image_lookup = list()
+		subscribed_mobs = list()
+
+	proc/add_image(image/img) // adds to the list and adds to all mobs clients directly
+		src.images.Add(img)
+		if (loc_to_image_lookup[img.loc])
+			loc_to_image_lookup[img.loc] += img
+		else
+			loc_to_image_lookup[img.loc] = list(img)
+			RegisterSignal(img.loc, COMSIG_MOB_PROP_INVISIBILITY, .proc/on_mob_invisibility_changed)
+
+		for (var/mob/iterated_mob as() in subscribed_mobs)
```

Comment:
what happens if a mob gets deleted - the entry will probably be null

Reply:
A disposing signal is registered on the mob when it's added to handle gracefully removing it from the list on deletion.